Adding local API to solution for tutorial purposes

saraford opened the pull request:
This PR adds a local ASP.NET Web API 2 to this WPF sample, similar to the [webapp-and-webapi sample](https://github.com/Azure-Samples/active-directory-b2c-dotnet-webapp-and-webapi), that uses the B2C demo tenant. Having both projects in the same solution will help us write self-contained getting started tutorials. Otherwise, if we used an existing or new API from a different repo, we'd have to keep track of how those changes impact this tutorial, hence my goal of keeping all getting started info within this one repo.

saraford committed:
added new local API resource

saraford committed:
added new Hello route

saraford committed:
enabled SSL

saraford committed:
added required nuget packages

saraford committed:
wired up the OWIN middleware

saraford committed:
only authorized users with proper scopes can access hello endpoint

saraford committed:
wpf app now calls the local Hello API

spottedmahn commented:
My notes for testing the PR: [OneNote Public Share Notebook (Read only)](https://1drv.ms/o/s!AmsBn7YVdmXPhdxV0jzOVTwIiLEWIw)

saraford commented:
Hey @spottedmahn :wave:

Thanks for the feedback! 🙏 Looks like the API call works as expected! 😄 That's all I changed in this PR. I tried to make the commits as short as possible to help indicate what was changing here.

For your other feedback: 
1. I've confirmed that these repro in master, so I've opened Issues to track: #10 and #11 
2. It looks like this sample app may have been designed to require signing-in when editing profile and after changing password. It was written before my time. Maybe we could show these scenarios in a separate branch (or vice versa)? If you'd like us to show these scenarios without resigning in, lmk and i'll open an Issue to track. 

Thanks again for the feedback! ✨

parakhj commented:
I downloaded the PR, ran the desktop app but not the api, signed in, and then called the api (even though the api wasn't running). Result shown below. You should add a try catch block somewhere to provide a better error message.

<!DOCTYPE html>
<html>
    <head>
        <title>Configuration Error</title>
        <meta name="viewport" content="width=device-width" />
        <style>
         body {font-family:"Verdana";font-weight:normal;font-size: .7em;color:black;} 
         p {font-family:"Verdana";font-weight:normal;color:black;margin-top: -5px}
         b {font-family:"Verdana";font-weight:bold;color:black;margin-top: -5px}
         H1 { font-family:"Verdana";font-weight:normal;font-size:18pt;color:red }
         H2 { font-family:"Verdana";font-weight:normal;font-size:14pt;color:maroon }
         pre {font-family:"Consolas","Lucida Console",Monospace;font-size:11pt;margin:0;padding:0.5em;line-height:14pt}
         .marker {font-weight: bold; color: black;text-decoration: none;}
         .version {color: gray;}
         .error {margin-bottom: 10px;}
         .expandable { text-decoration:underline; font-weight:bold; color:navy; cursor:hand; }
         @media screen and (max-width: 639px) {
          pre { width: 440px; overflow: auto; white-space: pre-wrap; word-wrap: break-word; }
         }
         @media screen and (max-width: 479px) {
          pre { width: 280px; }
         }
        </style>
    </head>

    <body bgcolor="white">

            <span><H1>Server Error in '/' Application.<hr width=100% size=1 color=silver></H1>

            <h2> <i>Configuration Error</i> </h2></span>

            <font face="Arial, Helvetica, Geneva, SunSans-Regular, sans-serif ">

            <b> Description: </b>An error occurred during the processing of a configuration file required to service this request. Please review the specific error details below and modify your configuration file appropriately.
            <br><br>

            <b> Parser Error Message: </b>The CodeDom provider type &quot;Microsoft.CodeDom.Providers.DotNetCompilerPlatform.CSharpCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=1.0.7.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35&quot; could not be located.<br><br>

            <b>Source Error:</b> <br><br>

            <table width=100% bgcolor="#ffffcc">
               <tr>
                  <td>
                      <code><pre>

Line 65:   &lt;system.codedom&gt;
Line 66:     &lt;compilers&gt;
<font color=red>Line 67:       &lt;compiler language=&quot;c#;cs;csharp&quot; extension=&quot;.cs&quot; type=&quot;Microsoft.CodeDom.Providers.DotNetCompilerPlatform.CSharpCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=1.0.7.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35&quot; warningLevel=&quot;4&quot; compilerOptions=&quot;/langversion:default /nowarn:1659;1699;1701&quot; /&gt;
</font>Line 68:       &lt;compiler language=&quot;vb;vbs;visualbasic;vbscript&quot; extension=&quot;.vb&quot; type=&quot;Microsoft.CodeDom.Providers.DotNetCompilerPlatform.VBCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=1.0.7.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35&quot; warningLevel=&quot;4&quot; compilerOptions=&quot;/langversion:default /nowarn:41008 /define:_MYTYPE=\&amp;quot;Web\&amp;quot; /optionInfer+&quot; /&gt;
Line 69:     &lt;/compilers&gt;</pre></code>

                  </td>
               </tr>
            </table>

            <br>

            <b> Source File: </b> C:\Users\parja\Source\Repos\active-directory-b2c-dotnet-desktop\HelloAPI-server\web.config<b> &nbsp;&nbsp; Line: </b> 67
            <br><br>

            <hr width=100% size=1 color=silver>

            <b>Version Information:</b>&nbsp;Microsoft .NET Framework Version:4.0.30319; ASP.NET Version:4.7.2552.0

            </font>

    </body>
</html>
<!-- 
[ConfigurationErrorsException]: The CodeDom provider type &quot;Microsoft.CodeDom.Providers.DotNetCompilerPlatform.CSharpCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=1.0.7.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35&quot; could not be located. (C:\Users\parja\Source\Repos\active-directory-b2c-dotnet-desktop\HelloAPI-server\web.config line 67)
   at System.CodeDom.Compiler.CompilerInfo.get_CodeDomProviderType()
   at System.Web.Configuration.CompilationSection.GetCompilerInfoFromLanguage(String language)
   at System.Web.UI.TemplateParser.ProcessLanguageAttribute(String language)
   at System.Web.UI.TemplateParser.ProcessMainDirectiveAttribute(String deviceName, String name, String value, IDictionary parseData)
   at System.Web.UI.TemplateParser.ProcessMainDirective(IDictionary mainDirective)
[HttpParseException]: The CodeDom provider type &quot;Microsoft.CodeDom.Providers.DotNetCompilerPlatform.CSharpCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=1.0.7.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35&quot; could not be located. (C:\Users\parja\Source\Repos\active-directory-b2c-dotnet-desktop\HelloAPI-server\web.config line 67)
   at System.Web.UI.TemplateParser.ProcessException(Exception ex)
   at System.Web.UI.TemplateParser.ParseStringInternal(String text, Encoding fileEncoding)
   at System.Web.UI.TemplateParser.ParseString(String text, VirtualPath virtualPath, Encoding fileEncoding)
   at System.Web.UI.TemplateParser.ParseFile(String physicalPath, VirtualPath virtualPath)
   at System.Web.UI.TemplateParser.ParseInternal()
   at System.Web.UI.TemplateParser.Parse()
   at System.Web.Compilation.BaseTemplateBuildProvider.get_CodeCompilerType()
   at System.Web.Compilation.BuildProvider.GetCompilerTypeFromBuildProvider(BuildProvider buildProvider)
   at System.Web.Compilation.BuildProvidersCompiler.ProcessBuildProviders()
   at System.Web.Compilation.BuildProvidersCompiler.PerformBuild()
   at System.Web.Compilation.ApplicationBuildProvider.GetGlobalAsaxBuildResult(Boolean isPrecompiledApp)
   at System.Web.Compilation.BuildManager.CompileGlobalAsax()
   at System.Web.Compilation.BuildManager.EnsureTopLevelFilesCompiled()
[HttpException]: The CodeDom provider type &quot;Microsoft.CodeDom.Providers.DotNetCompilerPlatform.CSharpCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=1.0.7.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35&quot; could not be located. (C:\Users\parja\Source\Repos\active-directory-b2c-dotnet-desktop\HelloAPI-server\web.config line 67)
   at System.Web.Compilation.BuildManager.ReportTopLevelCompilationException()
   at System.Web.Compilation.BuildManager.EnsureTopLevelFilesCompiled()
   at System.Web.Compilation.BuildManager.CallAppInitializeMethod()
   at System.Web.Hosting.HostingEnvironment.Initialize(ApplicationManager appManager, IApplicationHost appHost, IConfigMapPathFactory configMapPathFactory, HostingEnvironmentParameters hostingParameters, PolicyLevel policyLevel, Exception appDomainCreationException)
[HttpException]: The CodeDom provider type &quot;Microsoft.CodeDom.Providers.DotNetCompilerPlatform.CSharpCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=1.0.7.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35&quot; could not be located. (C:\Users\parja\Source\Repos\active-directory-b2c-dotnet-desktop\HelloAPI-server\web.config line 67)
   at System.Web.HttpRuntime.FirstRequestInit(HttpContext context)
   at System.Web.HttpRuntime.EnsureFirstRequestInit(HttpContext context)
   at System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest wr, HttpContext context)
-->

parakhj commented:
Not as important of a rough edge, but I called Edit Profile, got to the edit profile page, and then cancelled. Got the error below - could you get rid of the first line there?

Session has expired, please sign out and back in.https://login.microsoftonline.com/tfp/fabrikamb2c.onmicrosoft.com/b2c_1_edit_profile/oauth2/v2.0/authorize
Microsoft.Identity.Client.MsalServiceException: AADB2C90091: The user has cancelled entering self-asserted information.
Correlation ID: 6cf338d3-32a6-47bc-88df-1dffbe311688
Timestamp: 2017-11-17 23:45:16Z
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.VerifyAuthorizationResult()
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.<PreTokenRequest>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.<RunAsync>d__33.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.PublicClientApplication.<AcquireTokenForUserCommonAsync>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.PublicClientApplication.<AcquireTokenAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at active_directory_b2c_wpf.MainWindow.<EditProfileButton_Click>d__2.MoveNext() in C:\Users\parja\Source\Repos\active-directory-b2c-dotnet-desktop\active-directory-b2c-wpf\MainWindow.xaml.cs:line 70
	ErrorCode: access_denied
	StatusCode: 0
	Claims:

saraford committed:
using new API app portal registration

parakhj commented on HelloAPI-server/Web.config:
```diff
@@ -0,0 +1,71 @@
+    <add key="webpages:Enabled" value="false" />
+    <add key="ClientValidationEnabled" value="true" />
+    <add key="UnobtrusiveJavaScriptEnabled" value="true" />
+    <add key="ida:AadInstance" value="https://login.microsoftonline.com/{0}/v2.0/.well-known/openid-configuration?p={1}" />
```
Could you add spacing/comment to the relevant parts in this file so that it is easier for someone to know what to edit.

saraford replied:
:+1: Added a comment

parakhj commented on HelloAPI-server/Controllers/HelloController.cs:
```diff
@@ -0,0 +1,41 @@
+        public const string scopeElement = "http://schemas.microsoft.com/identity/claims/scope";
+        public const string objectIdElement = "http://schemas.microsoft.com/identity/claims/objectidentifier";
+
+        [Authorize]
```
Not that it makes any difference functionality wise, I would move this attribute at the class level. More clearly indicates that any call to this API will require authentication first (not just GET /hello)

saraford replied:
:+1: moved the attribute to the class level

parakhj commented on HelloAPI-server/Controllers/ValuesController.cs:
```diff
@@ -0,0 +1,39 @@
+using System.Web.Http;
+
+namespace HelloAPI_server.Controllers
+{
```
What is this class for?

saraford replied:
Deleted. It was created as part of the File - New - Project template.

parakhj commented on HelloAPI-server/Views/Home/Index.cshtml:
```diff
@@ -0,0 +1,24 @@
+﻿<div class="jumbotron">
+    <h1>ASP.NET</h1>
```
Could you add something here that indicates that this is an API.

saraford replied:
:+1: Updated with a couple of sentences. Just an 💡  we could add more instructions needed, like how to setup multiple projects if only the web API is running.

saraford committed:
added exception handling when API is not running

saraford committed:
removed non-applicable error message text

saraford committed:
added comments where to change tenant values for tutorial

saraford committed:
moved Authorize attribute to class level

saraford committed:
deleted default ValuesController.cs

saraford committed:
updated API homepage

updated to explain this is the API and which endpoint to use for the sample

saraford commented:
> and then called the api (even though the api wasn't running). Result shown below. You should add a try catch block somewhere to provide a better error message.

I added logic for this particular scenario. We need to address better error handling in all our samples. Let's create a separate issue to track improved error handling and feedback. 

> could you get rid of the first line there "Session has expired, please sign out and back "

Removed. This is another UX scenario we should address in all of our samples.

bgavrilMS commented:
This has been open for too long now. Please see [Microsoft.Identity.Web](https://github.com/AzureAD/microsoft-identity-web) for web api scenarios.
